### PR TITLE
Reimplement AutoAssign

### PR DESCRIPTION
**Pull request by jshilong**
I reimplemented autoassign.
I got a similar accuracy of 40.4 vs 40.5 in origin implementation, and the performance is unstable with  1x setting and may fluctuate by about 0.3 mAP. Results between 40.3 and 40.6 are acceptable.

### CODE REVIEW

File: mmdet/models/dense_heads/autoassign_head.py [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+            iou_function_max - iou_function_min).clamp(min=EPS)
+
+        foreground_idxs = torch.nonzero(inside_gt_bbox_mask, as_tuple=True)
+        neg_weight = neg_weight[inside_gt_bbox_mask]
+        p_neg_weight[foreground_idxs[0],
+                     gt_labels[foreground_idxs[1]]] = neg_weight
+
+        logits = (joint_conf * p_neg_weight)
+        neg_loss = (
+            logits**2 * F.binary_cross_entropy(
```
Comment (Johnson-Wang) [inline mmdet/models/dense_heads/autoassign_head.py:223]: Can we use focal loss interface?

File: mmdet/models/dense_heads/autoassign_head.py [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+
+        Args:
+            cls_score (Tensor): All category scores for each point on
+                the feature map. The shape is (num_points, num_class).
+            objectness (Tensor): Foreground probability of all points,
+                has shape (num_points, 1).
+            reg_loss (Tensor): The regression loss of each gt_bbox and each
+                prediction box, has shape of (num_points, num_gt).
+            gt_labels (Tensor): The zeros based gt_labels of all gt
+                with shape of (num_gt,).
```
[the remark this answers is not in the document]
Comment (jshilong) [reply mmdet/models/dense_heads/autoassign_head.py:161]: It's a bit counter-intuitive, this is num_gt in autossignhead, we do nothing with classification targets in `get_targets` fuction

File: mmdet/models/dense_heads/autoassign_head.py (hunk 1 above)
Comment (jshilong) [reply mmdet/models/dense_heads/autoassign_head.py:223]: Because logits have been sigmoid calculation  if we want to use focal loss, we have to convert it to before sigmoid, there will
be a meaningless calculation.

File: mmdet/models/dense_heads/autoassign_head.py [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+        Returns:
+            tuple[Tensor]:
+
+                - neg_loss (Tensor): The negative loss of all points \
+                    in the feature map.
+        """
+        num_gts = len(gt_labels)
+        joint_conf = (cls_score * objectness)
+        p_neg_weight = torch.ones_like(joint_conf)
+        if num_gts:
```
Comment (Johnson-Wang) [inline mmdet/models/dense_heads/autoassign_head.py:250]: It's better to use explicit expression `if num_gts > 0`.

**Review by Johnson-Wang — COMMENTED** [review]
General comments:
The overall code structure is fine to me despite too many hard-coded hyperparameters. Is it possible to specify `loss_pos` and `loss_neg` in `AutoAssignHead`, so that others may be free to change to other loss styles?

File: configs/autoassign/autoassign_r50_fpn_8x2_1x_coco.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    ),
+    bbox_head=dict(
+        type='AutoAssignHead',
+        norm_on_bbox=True,
+        num_classes=80,
+        in_channels=256,
+        stacked_convs=4,
+        feat_channels=256,
+        strides=[8, 16, 32, 64, 128],
+        loss_bbox=dict(type='GIoULoss', loss_weight=1.0),
```
Comment (ZwwWayne) [inline configs/autoassign/autoassign_r50_fpn_8x2_1x_coco.py:35]: clean comma

File: configs/autoassign/autoassign_r50_fpn_8x2_1x_coco.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        style='caffe'),
+    neck=dict(
+        type='FPN',
+        in_channels=[256, 512, 1024, 2048],
+        out_channels=256,
+        start_level=1,
+        add_extra_convs=True,
+        extra_convs_on_inputs=True,
+        num_outs=5,
+        relu_before_extra_convs=True,
```
Comment (ZwwWayne) [inline configs/autoassign/autoassign_r50_fpn_8x2_1x_coco.py:25]: clean comma

File: configs/autoassign/README.md [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+## Results and Models
+
+| Backbone  | Style   | Lr schd | Mem (GB) |   box AP | Config | Download |
+|:---------:|:-------:|:-------:|:--------:|:------:|:------:|:--------:|
+| R-50     | pytorch | 1x      | 4.08      |   40.4  |
+
+**Note**:
+
+1. We find that the performance is unstable with 1x setting and may fluctuate by about 0.3 mAP. We report the best results.
```
Comment (ZwwWayne) [inline configs/autoassign/README.md:11]: Suggest reporting the results that are close to the average.

File: mmdet/models/dense_heads/autoassign_head.py [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+        self.neg_loss_weight = neg_loss_weight
+        self.center_loss_weight = center_loss_weight
+
+    def init_weights(self):
+        """Initialize weights of the head.
+
+        In particular, we have special initialization for classified conv's and
+        regression conv's bias
+        """
+
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/autoassign_head.py:164]: [Nit]: may clean this empty line

File: mmdet/models/dense_heads/autoassign_head.py [hunk 8]
```diff
@@ -0,0 +1,475 @@
...
+    def _get_points_single(self,
+                           featmap_size,
+                           stride,
+                           dtype,
+                           device,
+                           flatten=False):
+        """Almost the same as the implementation in fcos, we remove half stride
+        offset to align with the original implementation."""
+
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/autoassign_head.py:178]: [Nit]: may clean this empty line

File: mmdet/models/dense_heads/autoassign_head.py [hunk 9]
```diff
@@ -0,0 +1,475 @@
...
+                of (num_points, num_gt). Each value represents
+                the center weighting coefficient.
+
+        Returns:
+            tuple[Tensor]:
+
+                - pos_loss (Tensor): The positive loss of all points
+                  in the gt_bboxes.
+        """
+
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/autoassign_head.py:209]: [Nit]: may clean this empty line

File: mmdet/models/dense_heads/autoassign_head.py [hunk 10]
```diff
@@ -0,0 +1,475 @@
...
+        num_levels = len(points)
+        concat_points = torch.cat(points, dim=0)
+        # the number of points per img, per lvl
+        num_points = [center.size(0) for center in points]
+        inside_gt_bbox_mask_list, bbox_targets_list = multi_apply(
+            self._get_target_single,
+            gt_bboxes_list,
+            points=concat_points,
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/autoassign_head.py:423]: clean comma

File: mmdet/models/dense_heads/autoassign_head.py [hunk 11]
```diff
@@ -0,0 +1,475 @@
...
+        )
+        bbox_targets_list = [
+            list(bbox_targets.split(num_points, 0))
+            for bbox_targets in bbox_targets_list
+        ]
+        if self.norm_on_bbox:
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/autoassign_head.py:429]: Can we do this in _get_target_single to simplify the logics?

**Comment by ZwwWayne** [comment]
1. Please resolve the conflicts and comments.
2. After resolving the comments, @jshilong can benchmark models for release

File: tests/test_models/test_heads.py [hunk 12]
```diff
@@ -4,16 +4,103 @@
...
+    }]
+    train_cfg = mmcv.Config(
+        dict(assigner=None, allowed_border=-1, pos_weight=-1, debug=False))
+    self = AutoAssignHead(
+        num_classes=4,
+        in_channels=1,
+        train_cfg=train_cfg,
+        loss_cls=dict(
+            type='CrossEntropyLoss', use_sigmoid=True, loss_weight=1.0),
+        loss_bbox=dict(type='GIoULoss', loss_weight=1.3),
```
Comment (ZwwWayne) [inline tests/test_models/test_heads.py:34]: clean comma.

**Comment by ZwwWayne** [comment]
Need to resolve conflicts before the merge.

File: mmdet/models/dense_heads/autoassign_head.py (hunk 11 above)
Comment (jshilong) [reply mmdet/models/dense_heads/autoassign_head.py:429]: In _get_target_single，tensor has a shape (num_points_all_levels, 4), we have to do this after line 426 which split the tensor to each level.

File: setup.cfg [hunk 13]
```diff
@@ -1,13 +0,0 @@
-[isort]
```
Comment (ZwwWayne) [inline setup.cfg:1]: This file should not be deleted.

File: configs/autoassign/autoassign_r50_fpn_8x2_1x_coco.py [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+    bbox_head=dict(
+        type='AutoAssignHead',
+        norm_on_bbox=True,
+        num_classes=80,
+        in_channels=256,
+        stacked_convs=4,
+        feat_channels=256,
+        strides=[8, 16, 32, 64, 128],
+        loss_bbox=dict(type='GIoULoss', loss_weight=5.0)))
+train_cfg = None
```
Comment (ZwwWayne) [inline configs/autoassign/autoassign_r50_fpn_8x2_1x_coco.py:35]: Move train_cfg/test_cfg into model

File: configs/autoassign/autoassign_r50_fpn_8x2_1x_coco.py [hunk 15]
```diff
@@ -0,0 +1,82 @@
+_base_ = [
+    '../_base_/datasets/coco_detection.py',
+    '../_base_/schedules/schedule_1x.py', '../_base_/default_runtime.py'
+]
+model = dict(
+    type='AutoAssign',
+    pretrained='open-mmlab://detectron/resnet50_caffe',
```
Comment (ZwwWayne) [inline configs/autoassign/autoassign_r50_fpn_8x2_1x_coco.py:7]: 1. If we use caffe model, we should indicate that in the config.
2. Detectron caffe is deprecated, we should use detectron2/resnet50_caffe

File: configs/autoassign/autoassign_r50_fpn_8x2_1x_coco.py [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+            dict(type='ImageToTensor', keys=['img']),
+            dict(type='Collect', keys=['img'])
+        ])
+]
+data = dict(
+    train=dict(pipeline=train_pipeline, filter_only_ignore=True),
+    val=dict(pipeline=test_pipeline),
+    test=dict(pipeline=test_pipeline))
+# optimizer
+optimizer = dict(lr=0.01, paramwise_cfg=dict(norm_decay_mult=0.))
```
Comment (ZwwWayne) [inline configs/autoassign/autoassign_r50_fpn_8x2_1x_coco.py:74]: check whether paramwise_cfg is necessary.

Comment (jshilong) [reply configs/autoassign/autoassign_r50_fpn_8x2_1x_coco.py:74]: This follows the original implementation.

File: mmdet/models/dense_heads/autoassign_head.py [hunk 17]
```diff
@@ -0,0 +1,492 @@
...
+
+    Args:
+        force_topk (bool): When no point falls into gt_bbox, forcibly
+            select the k points closest to the center to calculate
+            the center prior. Defaults to False.
+        topk (int): The number of points used to calculate the
+            center prior when no point falls in gt_bbox. Only work when
+            force_topk if True. Defaults to 9.
+        num_classes (int): The class number of dataset. Defaults to 80.
+        stride (tuple[int]): The stride of each input feature map. Defaults
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/autoassign_head.py:29]: stride -> strides

File: mmdet/models/dense_heads/autoassign_head.py [hunk 18]
```diff
@@ -0,0 +1,492 @@
...
+        """
+        inside_gt_bbox_mask = inside_gt_bbox_mask.clone()
+        num_gts = len(labels)
+        num_points = sum([len(item) for item in anchor_points_list])
+        if num_gts == 0:
+            return gt_bboxes.new_zeros(num_points,
+                                       num_gts), inside_gt_bbox_mask
+        center_prior_list = []
+        for slvl_points, stride in zip(anchor_points_list, self.strides):
+            single_level_points = slvl_points[:, None, :].expand(
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/autoassign_head.py:82]: Need to add comment to indicate what `slvl_points` stands for

File: mmdet/models/dense_heads/autoassign_head.py [hunk 19]
```diff
@@ -0,0 +1,492 @@
...
+                the center weighting coefficient.
+
+        Returns:
+            tuple[Tensor]:
+
+                - pos_loss (Tensor): The positive loss of all points
+                  in the gt_bboxes.
+        """
+
+        p_loc = torch.exp(-reg_loss)
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/autoassign_head.py:212]: Indicate what p_loc, p_cls, p_pos means, e.g., if they have similar names in the original paper or formula, we should indicate that in the comments.

File: mmdet/models/dense_heads/autoassign_head.py [hunk 20]
```diff
@@ -0,0 +1,492 @@
...
+        p_cls = (cls_score * objectness)[:, gt_labels]
+        p_pos = p_cls * p_loc
+
+        # 3 is a hyper-parameter to control the contributions of high and
+        # low confidence locations towards positive losses.
+        confidence_weight = torch.exp(p_pos * 3)
+        p_pos_weight = (confidence_weight * center_prior_weights) / (
+            (confidence_weight * center_prior_weights).sum(
+                0, keepdim=True)).clamp(min=EPS)
+        reweight_p_pos = (p_pos * p_pos_weight).sum(0)
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/autoassign_head.py:222]: reweight_p_pos -> reweighted_p_pos

File: mmdet/models/dense_heads/autoassign_head.py [hunk 21]
```diff
@@ -0,0 +1,492 @@
...
+                weight=None,
+                reduction_override='none')
+            reg_loss_list.append(loss_bbox.reshape(num_points, temp_num_gt))
+
+        cls_scores = [item.sigmoid() for item in cls_scores]
+        objectnesses = [item.sigmoid() for item in objectnesses]
+        pos_loss_list, = multi_apply(self.get_pos_loss_single, cls_scores,
+                                     objectnesses, reg_loss_list, gt_labels,
+                                     center_prior_weight_list)
+        pos_avg_factor = reduce_mean(bbox_pred.new_tensor(all_num_gt))
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/autoassign_head.py:377]: should we add `clamp_(min=1)` here

File: mmdet/models/dense_heads/autoassign_head.py [hunk 22]
```diff
@@ -0,0 +1,492 @@
...
+        pos_loss = sum(pos_loss_list) / pos_avg_factor
+
+        neg_loss_list, = multi_apply(self.get_neg_loss_single, cls_scores,
+                                     objectnesses, gt_labels, ious_list,
+                                     inside_gt_bbox_mask_list)
+        neg_avg_factor = sum(item.data.sum()
+                             for item in center_prior_weight_list)
+        neg_avg_factor = reduce_mean(neg_avg_factor)
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/autoassign_head.py:385]: Use `neg_avg_factor = reduce_mean(neg_avg_factor).clamp_(min=1)` do not use `max(neg_avg_factor, 1)`

**Review by ZwwWayne — APPROVED** [review]
